CesiumGltf/include/CesiumGltf/SharedAssetDepot.h

```diff
@@ -0,0 +1,260 @@
+   * @param outAverageAge The average time in seconds that the current deletion
+   * candidates have spent pending deletion.
+   * @param outCount The number of current deletion candidates.
+   */
+  void getDeletionStats(float& outAverageAge, size_t& outCount) const {
+    size_t count = 0;
+    float total = 0;
+
+    std::lock_guard lock(this->deletionCandidatesMutex);
+    for (auto& [asset, age] : this->deletionCandidates) {
+      total += age;
+      count++;
+    }
+
+    outAverageAge = count > 0 ? total / count : 0;
+    outCount = count;
+  }
+
+private:
+  // Disable copy
+  void operator=(const SharedAssetDepot<AssetType>& other) = delete;
+
+  /**
+   * Marks the given asset as a candidate for deletion.
+   * Should only be called by {@link SharedAsset}.
+   */
+  void markDeletionCandidate(
+      const CesiumUtility::IntrusivePointer<AssetType>& pAsset) {
+    std::lock_guard lock(this->deletionCandidatesMutex);
+    this->deletionCandidates.push_back(pAsset);
+    this->totalDeletionCandidateMemoryUsage += pAsset->getSizeBytes();
+
+    if (this->totalDeletionCandidateMemoryUsage > this->staleAssetSizeLimit) {
+      std::lock_guard assetsLock(this->assetsMutex);
+      while (!this->deletionCandidates.empty() &&
+             this->totalDeletionCandidateMemoryUsage >
+                 this->staleAssetSizeLimit) {
+        const CesiumUtility::IntrusivePointer<AssetType>& pOldAsset =
+            this->deletionCandidates.front();
+        this->deletionCandidates.pop_front();
+        this->assets.erase(pOldAsset->getUniqueAssetId());
+        this->totalDeletionCandidateMemoryUsage -= pOldAsset->getSizeBytes();
+      }
+    }
+  }
+
+  /**
+   * Unmarks the given asset as a candidate for deletion.
+   * Should only be called by {@link SharedAsset}.
+   */
+  void unmarkDeletionCandidate(
+      const CesiumUtility::IntrusivePointer<AssetType>& pAsset) {
+    std::lock_guard lock(this->deletionCandidatesMutex);
+    const std::string& assetId = pAsset->getUniqueAssetId();
+    for (auto it = this->deletionCandidates.begin();
+         it != this->deletionCandidates.end();
+         ++it) {
+      if ((*it)->getUniqueAssetId() == assetId) {
+        this->deletionCandidates.erase(it);
+        this->totalDeletionCandidateMemoryUsage -= (*it)->getSizeBytes();
+        break;
+      }
+    }
+  }
+
+  // Assets that have a unique ID that can be used to de-duplicate them.
+  std::unordered_map<std::string, CesiumUtility::IntrusivePointer<AssetType>>
+      assets;
+  // Futures for assets that still aren't loaded yet.
+  std::unordered_map<
+      std::string,
+      CesiumAsync::SharedFuture<CesiumUtility::IntrusivePointer<AssetType>>>
+      pendingAssets;
+  // Mutex for checking or editing the pendingAssets map
+  std::mutex assetsMutex;
+
+  // List of assets that are being considered for deletion, in the order that
+  // they were added.
+  std::list<CesiumUtility::IntrusivePointer<AssetType>> deletionCandidates;
```

Comment:
It would be better to use our `DoublyLinkedList` class instead of `std::list`. Because with `DoublyLinkedList` the data item itself (e.g., `ImageCesium`) is the linked list node, there would be no need to search for it. So `unmarkDeletionCandidate` would become O(1) instead of O(n) (where n is the number of deletion candidates). It also avoids the need to new/delete linked list node instances. The down side is that the previous/next pointers need to be embedded in the asset class, but I think that's a reasonable tradeoff here. See [Tile](https://github.com/CesiumGS/cesium-native/blob/v0.40.1/Cesium3DTilesSelection/include/Cesium3DTilesSelection/Tile.h#L557) for an example of how to set this up.